### ISSUE TITLE
StrictNullChecks and throw/return with incorrect when used with closure scope

### ISSUE THREAD

**Issue by xealot** [Microsoft/TypeScript#13572]
**TypeScript Version:**  2.1.5

**Code**

```ts
function other() {
  return new Promise((resolve, reject) => {
    resolve("OK!");
  });
}

function test(hosts: {whatever: boolean}[]) {
  let host = hosts.filter((f) => f.whatever).shift();

  if (!host) {
    throw new Error("Crap!");
  }

  return other().then(() => {
    console.log(`${host.whatever} selected`);
  });
}
```
[Playground](http://www.typescriptlang.org/play/index.html#src=%0D%0Afunction%20other()%20%7B%0D%0A%20%20return%20new%20Promise((resolve%2C%20reject)%20%3D%3E%20%7B%0D%0A%20%20%20%20resolve(%22OK!%22)%3B%0D%0A%20%20%7D)%3B%0D%0A%7D%0D%0A%0D%0Afunction%20test(hosts%3A%20%7Bwhatever%3A%20boolean%7D%5B%5D)%20%7B%0D%0A%20%20let%20host%20%3D%20hosts.filter((f)%20%3D%3E%20f.whatever).shift()%3B%0D%0A%0D%0A%20%20if%20(!host)%20%7B%0D%0A%20%20%20%20throw%20new%20Error(%22Crap!%22)%3B%0D%0A%20%20%7D%0D%0A%0D%0A%20%20return%20other().then(()%20%3D%3E%20%7B%0D%0A%20%20%20%20console.log(%60%24%7Bhost.whatever%7D%20selected%60)%3B%0D%0A%20%20%7D)%3B%0D%0A%7D%0D%0A%0D%0A)

**Expected behavior:**

Do not emit an error about `host` possibly being null, since the method exits early if that's the case.

**Actual behavior:**

Typically when using `strictnullchecks` Typescript is intelligent enough to understand that when you throw an error from a method based on a type check/guard it narrows the type after the statement. As indicated in this short example:

```ts
function test(hosts: {whatever: boolean}[]) {
  let host = hosts.filter((f) => f.whatever).shift();

  // Correctly calls out hose as possibly being null
  console.log(`${host.whatever} selected`);
  
  if (!host) {
    throw new Error("Crap!");
  }

  // Does not complain here because of the above check/throw
  console.log(`${host.whatever} selected`);
}
```

For some reason when you add the return of the function and use `host` through a closure it starts complaining. 

I'm sure it's the closure, but I don't see how this could error.

**Comment by xealot** [comment]
Note: re-assigning the variable to a new variable after the throw seems to re-bind the type information and solves the issue... Seems like a bug.

```ts
function other() {
  return new Promise((resolve, reject) => {
    resolve("OK!");
  });
}

function test(hosts: {whatever: boolean}[]) {
  let host = hosts.filter((f) => f.whatever).shift();

  if (!host) {
    throw new Error("Crap!");
  }

  let ho = host;  // WAVES MAGIC WAND

  return other().then(() => {
    console.log(`${ho.whatever} selected`);
  });
}
```

**Comment by HerringtonDarkholme** [comment]
#9998 / #11498

**Comment by HerringtonDarkholme** [comment]
#9998 / #11498

Changing `let host` to `const host` also solves this.

**Comment by xealot** [comment]
Wow, the perfect comment!

Found the threads I knew existed by couldn't locate in search, *and* offered clean workaround!

Thanks. Closing in favor of #9998 and #11498
